client/src/main/java/io/pravega/client/stream/impl/StreamCutImpl.java

```diff
@@ -62,4 +84,68 @@ public boolean validate(Set<String> segmentNames) {
 
         return true;
     }
+    
+    private static class StreamCutBuilder implements ObjectBuilder<StreamCutInternal> {
+    }
+    
+    public static class StreamCutSerializer extends VersionedSerializer.WithBuilder<StreamCutInternal, StreamCutBuilder> {
+        @Override
+        protected StreamCutBuilder newBuilder() {
+            return builder();
+        }
+
+        @Override
+        protected byte getWriteVersion() {
+            return 0;
+        }
+
+        @Override
+        protected void declareVersions() {
+            version(0).revision(0, this::write00, this::read00);
+        }
+
+        private void read00(RevisionDataInput revisionDataInput, StreamCutBuilder builder) throws IOException {
+            Stream stream = Stream.of(revisionDataInput.readUTF());
+            builder.stream(stream);
+            Map<Segment, Long> map = revisionDataInput.readMap(in -> new Segment(stream.getScope(),
```

Comment:
So... You write a Map of Ints to Longs and read a Map of Segments to Longs? 

Are you inferring data for `Segment` from the `stream`? If so, where is this serialized? Is `Stream.of` looking up data externally or is it serialized in that String? 

If it's serialized in the string, then this is (mostly) ok.
If it's fetching data externally, then this serialization is not correct. All the information in the deserialization should be inferred solely from what has been serialized, and not external sources.

Reply:
This class contains a stream, and a map from the segments in that stream to the offsets in those segments. During the serialization process it is first serializing the stream, followed by the map, but the map is transformed to strip out the stream name from each key, and just store the segment number. I do this because this information is redundant (every key is a segment from the same stream), and It means I don't have to worry about the corrupt corner case where they aren't the same.